framework/src/problems/FEProblemBase.C

```diff
@@ -747,9 +747,13 @@ FEProblemBase::initialSetup()
       adaptivity().uniformRefineWithProjection();
     }
   }
-  else if (_mesh.uniformRefineLevel() > 0 && !_app.isUltimateMaster() && !_app.masterMesh())
-    mooseError(
-        "Doing extra refinements when recovering is NOT supported for sub-apps of a MultiApp");
+  else
+  {
+    if (_mesh.uniformRefineLevel() > 0 && _app.getExodusFileRestart())
+      if (!_app.isUltimateMaster() && !_app.masterMesh())
+        mooseError(
+            "Doing extra refinements when recovering is NOT supported for sub-apps of a MultiApp");
+  }
```

Comment:
If you are going to add this error, you should comment on why this is not supported. The above error is somewhat of a mystery; we do not want to add another mystery

Reply:
I believe the restart error was added in https://github.com/idaholab/moose/pull/5693 and @friedmud created an issue for it in https://github.com/idaholab/moose/issues/5691

Reply:
I think we must include this error. I added an inline comment to make clear what it is for.

Reply:
If we add the flexible restart capability for addressing #21011, I think we can remove the Exodus restart completely and this error.

Reply:
exodus file restart will never happen when recovering:
```c++
  if ((_app.isRestarting() || _app.isRecovering()) && (_app.isUltimateMaster() || _force_restart))
  {
    TIME_SECTION("restartFromFile", 3, "Restarting From File");

    _restart_io->readRestartableDataHeader(true);
    _restart_io->restartEquationSystemsObject();

    /**                                                                                                               
     * TODO: Move the RestartableDataIO call to reload data here. Only a few tests fail when doing                    
     * this now. Material Properties aren't sized properly at this point and fail across the board,                   
     * there are a few other misc tests that fail too.                                                                
     *                                                                                                                
     * _restart_io->readRestartableData();                                                                            
     */
  }
  else
  {
    ExodusII_IO * reader = _app.getExReaderForRestart();

    if (reader)
    {
      TIME_SECTION("copyingFromExodus", 3, "Copying Variables From Exodus");

      _nl->copyVars(*reader);
      _aux->copyVars(*reader);
    }
```

Reply:
My understanding of the code is this:

When either `_app.isRecovering()` or `_app.isRestarting()`, we are just going to write over the subapp solution vector when we restore the cached subapp backup; the restoration of the backup occurs after where we copy variables from exodus so it doesn't matter what the solution held by the exodus file was. What I think we should do instead is simply not try to read exodus solutions when we are a subapp and `_app.isRecovering() || _app.isRestarting()`

Reply:
Do you have specific code changes for this? I think I agree with you.

Reply:
If we can improve `libMesh::ExodusII_IO` two functions `copy_nodal_solution` and `copy_elemental_solution` to support a uniform refinement level argument, we will not have this mess.